Refactored Boards to MVVM

allanfang1 opened the pull request:
There's some lag opening a board vs a note

allanfang1 committed:
Refactored Boards to MVVM

There's some lag opening a board vs a note

allanfang1 committed:
Rename NoteViewModel

ally0102 commented on app/src/main/java/com/allan/boardbuddies/fragments/BoardsFragment.java:
```diff
@@ -90,36 +74,29 @@ protected void populateMemoViewHolder(MemoViewHolder holder, Board element) {
-                }
-            }
+        @Override
+        public boolean areContentsTheSame(Board oldItem, Board newItem) {
```
what's the difference between this and `areItemsTheSame` above?

allanfang1 replied:
areItemsTheSame - used to determine structural changes between old and new list (additions/removals/position changes)

areContentsTheSame - determines if particular item was updated

ally0102 commented on app/src/main/java/com/allan/boardbuddies/fragments/EditBoardFragment.java:
```diff
@@ -0,0 +1,116 @@
+        editBoardViewModel = new ViewModelProvider(requireActivity()).get(EditBoardViewModel.class);
+
+        ViewTreeObserver viewTreeObserver = canvasView.getViewTreeObserver();
+        if (viewTreeObserver.isAlive()) {
```
curious, when would viewTreeObserver not be alive?

would recommend putting everything after this point into `onViewCreated`, since this is all logic that operates on the created views. you should only be inflating views in `onCreateView`

allanfang1 replied:
I can probably not have it? It's not alive when the view has been destroyed or just missing... which i think shouldn't happen

allanfang1 replied:
Ok but like what if I never need to access the object again e.g: bind toolbar to view, init toolbar. 
Total 2 usages, but i'd have to make the toolbar a fragment property. Is it worth

ally0102 commented on app/src/main/java/com/allan/boardbuddies/fragments/EditBoardFragment.java:
```diff
@@ -0,0 +1,116 @@
+        BottomNavigationView bottomNavigationView = view.findViewById(R.id.board_bottom_nav);
+        bottomNavigationView.setOnItemSelectedListener(this);
+
+        getChildFragmentManager().setFragmentResultListener(Constants.TEXT_DIALOG_FRAGMENT_ADD_TEXT_REQUEST_KEY, this, new FragmentResultListener() {
+            @Override
+            public void onFragmentResult(@NonNull String requestKey, @NonNull Bundle bundle) {
+                String result = bundle.getString(Constants.TEXT_DIALOG_FRAGMENT_ADDED_TEXT);
+                String resultKey = bundle.getString("resultKey");
+                if ("add_text".equals(resultKey) && !result.isEmpty()){
+                    canvasView.newTextBox(result);
+                } else if ("edit_canvas_title".equals(resultKey)){
+                    boardTitle.setText(result);
+                }
+            }
+        });
```
as discussed, can we move this into the viewmodel? TextDialogFragment can just modify either the title of the board or add a new text, and it will reflect here

allanfang1 replied:
not in scope, next ticket

ally0102 commented on app/src/main/java/com/allan/boardbuddies/viewmodels/EditBoardViewModel.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.ArrayList;
+
+public class EditBoardViewModel extends ViewModel {
+    private Board localBoard;
```
LiveData?

allanfang1 replied:
oh true i'll fix it

allanfang1 committed:
Refactor dialogs to use viewmodel

MutableLiveData for strokes and texts is not necessary atm... might need it for later?

ally0102 commented on app/src/main/java/com/allan/boardbuddies/repositories/BoardRepository.java:
```diff
@@ -0,0 +1,83 @@
+        loadNotes(directory);
+    }
+
+    private void loadNotes(File directory){
```
nit: name this loadBoards?

ally0102 commented on app/src/main/java/com/allan/boardbuddies/repositories/BoardRepository.java:
```diff
@@ -0,0 +1,83 @@
+
+public class BoardRepository {
+    private static volatile BoardRepository boardRepository;
+
```
interesting...is this how it's done in java? to maintain a single instance for the entire application lifecycle?

allanfang1 replied:
Idk I just ended up using the singleton architecture. If you have a better recommendation we can talk

ally0102 commented on app/src/main/java/com/allan/boardbuddies/fragments/TitleDialogFragment.java:
```diff
@@ -0,0 +1,63 @@
+
+        Button button = view.findViewById(R.id.dialog_button);
+        button.setOnClickListener(v -> {
+            editBoardViewModel.getTitle().setValue(String.valueOf(captionText.getText()));
```
you should create a setter function in the viewmodel instead. viewmodel should only expose immutable live data that's read access only

ally0102 commented on app/src/main/java/com/allan/boardbuddies/viewmodels/EditBoardViewModel.java:
```diff
@@ -0,0 +1,57 @@
+        this.texts.setValue(new ArrayList<TextBox>(localBoard.getTexts().stream().map(TextBox::new).collect(Collectors.toList())));
+    }
+
+    public MutableLiveData<String> getTitle(){
```
see comment above - we should only be returning immutable live data

ally0102 commented on app/src/main/java/com/allan/boardbuddies/fragments/EditBoardFragment.java:
```diff
@@ -0,0 +1,114 @@
+    @Override
+    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
+        switch (item.getItemId()){
+            case R.id.add_text:
```
remind me again how you would go about adding a text box? don't you just click anywhere on the board and not the nav bar?

my understanding is that this callback would only handle the case where you click the nav bar, right?

allanfang1 replied:
No you get a new textbox if you click the navbar button

ally0102 reviewed (approved):
overall lgtm please address comments especially for MutableLiveData

allanfang1 committed:
minor pr fixes
